### PR TITLE
Add gateway for user signups statistics

### PR DESCRIPTION
**Pull request by naliwajek**
It allows to fetch signups for today and total number of
signups before end of today.

### CODE REVIEW

File: spec/lib/gateway/stat_gateway_spec.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
+describe StatGateway do
+  before do
+    DB[:userdetails].truncate
```
Comment (adrianclay) [inline spec/lib/gateway/stat_gateway_spec.rb]: Be nice to remove this duplication of truncate at some point, maybe using a truncation gem?

File: spec/lib/gateway/stat_gateway_spec.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    it 'returns zero signups for today' do
+      expect(subject.signups).to eq(today: 5, total: 6)
+    end
+  end
+
+  context 'given 2 signups tomorrow' do
+    before do
+      2.times do |i|
+        User.create(username: "Today #{i}", created_at: Date.today + 1)
```
Comment (adrianclay) [inline spec/lib/gateway/stat_gateway_spec.rb]: "Today #{i}" is a bit misleading for this testcase.

File: spec/lib/gateway/stat_gateway_spec.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  context 'given 5 signups today and 1 yesterday' do
+    before do
+      User.create(username: 'Yesterday', created_at: Date.today - 1)
+
+      5.times do |i|
+        User.create(username: "Today #{i}")
+      end
+    end
+
+    it 'returns zero signups for today' do
```
Comment (adrianclay) [inline spec/lib/gateway/stat_gateway_spec.rb]: ` it 'returns 5 signups for today and 6 in total' ` ?

File: spec/lib/gateway/stat_gateway_spec.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+  context 'given 3 signups today' do
+    before do
+      3.times do |i|
+        User.create(username: "Today #{i}")
+      end
+    end
+
+    it 'returns zero signups for today' do
```
Comment (adrianclay) [inline spec/lib/gateway/stat_gateway_spec.rb]: ` it 'returns 3 signups for today and 3 in total' ` ?

**Comment by adrianclay** [comment]
I'd add a bit of spiel to the commit message about using these stats and pushing them into the "performance platform".

**Comment by naliwajek** [comment]
@adrianclay none of those commits are doing any pushing, though.

What if this gateway is going to be used in the future for pushing stats to other systems? Are we going to go back to files and churn them to update theirs, no longer valid, commit message?

Separation of concerns on code level should correspond to commit messages as well IMO. If using the gateway for a new system doesn't force us to update it (that's the whole point of correct flow of control) then I don't see why we should include a message about upstream components' feature.

Use Case commit message will describe this and that should be a starting point for anyone who wants to know what app is doing.

File: spec/lib/gateway/stat_gateway_spec.rb [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    it 'counds them against SMS total signups' do
+      expect(subject.signups[:sms_total]).to eq(2)
+    end
+
+    it 'counds them against SMS today\'s signups' do
+      expect(subject.signups[:sms_today]).to eq(1)
+    end
+  end
+end
```
Comment (naliwajek) [inline spec/lib/gateway/stat_gateway_spec.rb]: Some of those tests will quickly become redundant and noisy once we add two last stats. I don't want to create a mega PR so I ought to merge it now and refactor once we have the whole gateway working.

File: spec/lib/gateway/stat_gateway_spec.rb (hunk 1 above)
Comment (naliwajek) [reply spec/lib/gateway/stat_gateway_spec.rb]: Agree.

File: spec/lib/gateway/stat_gateway_spec.rb [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    it 'counds them against SMS total signups' do
+      expect(subject.signups[:sms_total]).to eq(1)
+    end
+
+    it 'counds them against SMS today\'s signups' do
+      expect(subject.signups[:sms_today]).to eq(1)
+    end
+  end
+
+  context 'given SNS signups made on different dates' do
```
Comment (adrianclay) [inline spec/lib/gateway/stat_gateway_spec.rb]: Typo, in SNS.

File: spec/lib/gateway/stat_gateway_spec.rb [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        contact: '+0123456789',
+        sponsor: '+0123456789'
+        )
+    end
+
+    it 'counts them against total singups' do
+      expect(subject.signups[:total]).to eq(2)
+    end
+
+    it 'counds them against today\'s signups' do
```
Comment (adrianclay) [inline spec/lib/gateway/stat_gateway_spec.rb]: Type on counds.

Comment (adrianclay) [reply spec/lib/gateway/stat_gateway_spec.rb]: Also don't think you need the apostrophe, here, but don't trust my English.

**Comment by adrianclay** [comment]
I'd argue that context is super important, and that's what people reading commit messages from the past will find useful.

File: spec/lib/gateway/stat_gateway_spec.rb [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    it 'returns zero signups for today' do
+      expect(subject.signups[:today]).to eq(0)
+    end
+
+    it 'returns zero signups total' do
+      expect(subject.signups[:today]).to eq(0)
+    end
+  end
+
+  context 'given 1 SMS signup today' do
```
Comment (adrianclay) [inline spec/lib/gateway/stat_gateway_spec.rb]: Given 1 sms signup and 1 email signup?

I'd then also be explicit around setting the contact field.

File: spec/lib/gateway/stat_gateway_spec.rb (hunk 5 above)
Comment (adrianclay) [reply spec/lib/gateway/stat_gateway_spec.rb]: Sounds like a plan.

**Pull request merged**